### DISCUSSION TITLE
how to properly use  runOrigialID

### DISCUSSION THREAD

**Discussion by gsohler** [elalish/manifold#749] [Q&A]
Hi, I'd like to use runOriginalID' property for finding this info in the output when setting in the input.
When my input cube has 12 triangles, I supply 12 runOriginalID's
I have this code .:

 mesh.vertPos.resize(ps_tri->vertices.size());
 for(int i=0;i<ps_tri->vertices.size();i++) {
 Vector3d pt=ps_tri->vertices[i];
 mesh.vertPos[i] = glm::vec3((float) pt[0], (float) pt[1], (float) pt[2]);
 }
  
 mesh.triVerts.reserve(ps_tri->indices.size());
 for(int i=0;i<ps_tri->indices.size();i++) {
 auto &tri = ps_tri->indices[i];
 assert(tri.size() == 3);
 mesh.triVerts.emplace_back(tri[0], tri[1], tri[2]);
 }
  
 manifold::MeshGL meshgl(mesh);
 std::vector<unsigned int> originalID's;
  
 /* from Documention: at
 https://manifoldcad.org/docs/html/structmanifold_1_1_mesh_g_l.html
Optional: The OriginalID of the mesh this triangle run came from. This ID is ideal for reapplying materials to the output mesh. Multiple runs may have the same ID, e.g. representing different copies of the same input mesh. If you create an input [MeshGL](https://manifoldcad.org/docs/html/structmanifold_1_1_mesh_g_l.html) that you want to be able to reference as one or more originals, be sure to set unique values from ReserveIDs(). */  

 meshgl.runoriginalIDfaceID = originalID; 

---------------
Did some trials but I always get an vector out of range error.
I could not figure out, how to properly create values for the originalID's vector. ReserveID's creates weird output values depending on the input and even if i use them, the error persists.

Can I find sample code somewhere ?

**Comment by elalish** [comment]
If you're using C++, our tests are a good place to start. Likewise, if you run into a problem, writing a new `TEST` is always a welcome contribution. Take a look at our test helper functions from https://github.com/elalish/manifold/blob/master/test/manifold_test.cpp#L563

Your supplied code doesn't appear to assign any OriginalIDs at all? Also, I'd recommend using only `MeshGL`, rather than starting with `Mesh`. From your description it sounds like you may have conflated `OriginalID` (a mesh ID) with `faceID` (a triangle ID).

**Reply by gsohler** [reply]
Hi,
Thank you for your answer,
Basically I'd like to union a solid which already has red and green triangles which another solid , which has blue triangles and get a result with 3 differently colored triangles.
From documentation it appeared, that I have to set faceID's in the inputs to different values, but when I looked into the output, i only
found index locations to the source triangle but not my own data,
Is there any way to see my own data make it through the engine ?

**Reply by elalish** [reply]
Yes, there definitely is. Make your solids cubes or tetrahedra for now to keep things simple. Then post a fully-worked example code here and we can help you debug it. It may also be helpful to know how you're organizing your data outside of Manifold - what data structures do you use to keep track of these triangle colors for input and output? Or do you care yet?

**Comment by gsohler** [comment]
Hi Emmett,

I have coded a small code snippet, which demonstrates what I trying to do.
In here I union a red  cube and a blue cube  with overlapping corners.
In the result i hope to find red and blue triangles, but I just see just original face index locations.
Outside of manifold i have a vector with color codes and manifold shall only correctly maintain the index into this array.
Code is here:  https://bpa.st/N47Q
Hope that this is something, you can work with :)

Thank you!

**Reply by elalish** [reply]
Thanks, I see the problem now. The faceID really is meant for IDing faces, not materials. Since your materials are uniform across a given cube, `originalID` is a better solution here. You can make a map of `OriginalID -> color`, and then the output `MeshGL` will be sorted into two triangle runs, one per ID/color. 

If you want your inputs to have difference colors on different faces, you'll want to sort those into triangle runs with separate originalIDs in a similar way.

**Comment by gsohler** [comment]
Hi Emmett,
Thank you for your answer. I really want to process multi-colored input for
each operand of union to become versatile enough for
my intent. This means I want to start off with the last approach with
different colors on different faces.

My Problem right now is that I don't have any clue about the
orgiginalID concept you are referring to. supposedly you are referring to
the runOriginalID property and it appears that I can specify at most one
for each input ?
If I have 2 inputs to union and both are partially red and partially green,
does it mean that I need to use 4 different runOriginalID's ?
Will I have to run 2 x 2 runs like:
red union red
red union green
green union red
green union green ?

will manifold work correctly when only seeing red faces on 1st input which
is not manifold alone ?

Right now its rather hard to me see the correct strategy to union multi
colored inputs , basically I am lost. Would it be at least possible to add
sketches to my code and guide me to the right direction ?

Thank you for your attention.

**Reply by elalish** [reply]
I'm not going to write your C++ for you - there is far too much code around the Manifold code that could have problems already. I would start by simply doing this:
```
Manifold redCube = Manifold::cube(2);
Manifold greenCube = Manifold::cube(3).Translate({1,1,1});
Manifold union = redCube + greenCube;
MeshGL out = union.GetMeshGL();
```
Then dump every member of `out` to the command line and study it, along with our docs. Note the two `originalIDs`, corresponding in this case to red and green, as well as the associated triangle runs showing which portions of the manifold's triangles to use for each draw call. Once you understand that construction, you'll be able to create them on your own, since you can do `Manifold union2(out)` and `union2` and `union` will be identical. Did you study our `RelatedGL` test function? That will show you how to walk the triangle runs, and how they relate to input meshes.

**Comment by gsohler** [comment]
thank you for your answer. using information from faceID together with runIndex helped me to solve the problem.
I was not aware that "run" refers to process each input of the boolop individually and concating the runs effectivle yields the complete result.
